src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/Debugger/WebServer.cs

```diff
@@ -0,0 +1,1022 @@
+        {
+            var settings = _currentSettings;
+            if (settings == null)
+            {
+                return;
+            }
+
+            if (settings.ServerType == ServerType.IISExpress)
+            {
+                await _threadingService.SwitchToUIThread();
+
+                IVsDeveloperWebServer webServer = GetWebServerController();
+                if (webServer != null && webServer.IsRunning())
+                {
+                    webServer.Stop();
+                }
+            }
+        }
+
+        public async Task<(string exePath, string commandLine)> GetWebServerCommandLineAsync()
+        {
+            var settings = _currentSettings;
+
+            if (settings != null && settings.ServerType == ServerType.IISExpress)
+            {
+                await _threadingService.SwitchToUIThread();
+
+                IVsDeveloperWebServer webServer = GetWebServerController();
+                webServer.GetProcessInformation(out string exePath, out string commandLine);
+                return (exePath, commandLine);
+            }
+
+            return (string.Empty, string.Empty);
+        }
+
+        public async Task<int> GetWebServerProcessIdAsync()
+        {
+            var settings = _currentSettings;
+
+            if (settings != null && settings.ServerType == ServerType.IISExpress)
+            {
+                await _threadingService.SwitchToUIThread();
+
+                IVsDeveloperWebServer webServer = GetWebServerController();
+                webServer.GetExistingProcess(out uint pid);
+                return (int)pid;
+            }
+
+            return 0;
+        }
+
+        private string ProjectDirectory => Path.GetDirectoryName(_projectServices.Project.FullPath);
+        private string ProjectName => Path.GetFileNameWithoutExtension(_projectServices.Project.FullPath);
+        public string? TargetFrameworkMoniker
+        {
+            get
+            {
+                _projectVsServices.VsHierarchy.GetProperty(VsHierarchyPropID.TargetFrameworkMoniker, defaultValue: null, result: out string? targetFramework);
+                return targetFramework;
+            }
+        }
+
+        public Version TargetFrameworkVersion
+        {
+            get
+            {
+                var moniker = TargetFrameworkMoniker;
+                if (moniker != null)
+                {
+                    return new FrameworkName(moniker).Version;
+                }
+
+                return new Version();
+            }
+        }
+
+        public string? ProjectUrl
+        {
+            get
+            {
+                if (_currentSettings == null || _currentSettings.ServerUrls.Count == 0)
+                {
+                    return null;
+                }
+
+                return _currentSettings.ServerUrls[0];
+            }
+        }
+
+        public IReadOnlyList<string> WebServerUrls => _currentSettings?.ServerUrls ?? new List<string>();
```

Comment:
```suggestion
        public IReadOnlyList<string> WebServerUrls => _currentSettings?.ServerUrls ?? Array.Empty<string>();
```

Reply:
Ah, because it is a readonly list.  Didn't realize that